File: Web/backend/src/main/java/com/ByTech/ByTech/ShoppingCart/services/ShoppingCartService.java
package com.ByTech.ByTech.ShoppingCart.services;

import com.ByTech.ByTech.Message.Message;
import com.ByTech.ByTech.ShoppingCart.models.ShoppingCart;
import com.ByTech.ByTech.ShoppingCart.repositories.ShoppingCartRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;

@Service
@Transactional
public class ShoppingCartService {

    private final ShoppingCartRepository shoppingCartRepository;

    @Autowired
    public ShoppingCartService(ShoppingCartRepository shoppingCartRepository) {
        this.shoppingCartRepository = shoppingCartRepository;
    }

    public List<ShoppingCart> getListByClient(String userName){
        return this.shoppingCartRepository.findByClient_UserName(userName);
    }

    public ShoppingCart findByClientAndProduct(String userName, Long productId) {
        return this.shoppingCartRepository.findByClient_UserNameAndProduct_Id(userName, productId);
    }


    public void cleanShoppingCart(Long clientId){
        this.shoppingCartRepository.deleteByClient_Id(clientId);
    }

    public void removeProduct(Long id){
        this.shoppingCartRepository.deleteById(id);
    }

    public ResponseEntity<Message> addProduct(ShoppingCart shoppingCart) {
        ShoppingCart existingItem = findByClientAndProduct(shoppingCart.getClient().getUserName(), shoppingCart.getProduct().getId());
        
        if (existingItem != null) {
            existingItem.setAmount(existingItem.getAmount() + shoppingCart.getAmount());
            updateProduct(existingItem);
            System.out.println("Cantidad de producto actualizada: " + existingItem.getAmount());
            return new ResponseEntity<>(new Message("Cantidad de producto actualizada"), HttpStatus.OK);
        } else {
            addNewProduct(shoppingCart);
            System.out.println("Producto agregado al carrito: " + shoppingCart.getProduct().getName() + " - Cantidad: " + shoppingCart.getAmount());
            return new ResponseEntity<>(new Message("Producto agregado al carrito"), HttpStatus.OK);
        }
    }

    private void addNewProduct(ShoppingCart shoppingCart) {
        this.shoppingCartRepository.save(shoppingCart);
    }

    public void updateProduct(ShoppingCart shoppingCart){
        this.shoppingCartRepository.save(shoppingCart);
    }

    public Long getCountByClient(Long clientId){
        return this.shoppingCartRepository.countByClient_Id(clientId);
    }

    public ShoppingCart updateByID(ShoppingCart request, Long id ){
        ShoppingCart shoppingCart = shoppingCartRepository.findById(id).orElse(null);
        if(shoppingCart != null) {
            shoppingCart.setAmount(request.getAmount());
            shoppingCart.setClient(request.getClient());
            shoppingCart.setProduct(request.getProduct());

            shoppingCartRepository.save(shoppingCart);
        }
        return shoppingCart;
    }

    public ShoppingCart findById(Long id) {
        return shoppingCartRepository.findById(id).orElse(null);
    }
}

File: Movil/backend/ByTech_movil/src/main/kotlin/com/example/ByTech_movil/ShoppingCart/services/ShoppingCartService.kt
package com.example.ByTech_movil.ShoppingCart.services

import com.example.ByTech_movil.Message.Message
import com.example.ByTech_movil.ShoppingCart.models.ShoppingCart
import com.example.ByTech_movil.ShoppingCart.repositories.ShoppingCartRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.List

@Service
@Transactional
class ShoppingCartService @Autowired constructor(shoppingCartRepository: ShoppingCartRepository) {
    private val shoppingCartRepository: ShoppingCartRepository = shoppingCartRepository

    fun getListByClient(userName: String?): List<ShoppingCart> {
        return shoppingCartRepository.findByClient_UserName(userName)
    }

    fun findByClientAndProduct(userName: String?, productId: Long?): ShoppingCart? {
        return shoppingCartRepository.findByClient_UserNameAndProduct_Id(userName, productId)
    }

    fun cleanShoppingCart(clientId: Long?) {
        shoppingCartRepository.deleteByClient_Id(clientId)
    }

    fun removeProduct(id: Long?) {
        if (id != null) shoppingCartRepository.deleteById(id)
    }

    fun addProduct(shoppingCart: ShoppingCart): ResponseEntity<Message> {
        val existingItem = findByClientAndProduct(
            shoppingCart.client?.userName,
            shoppingCart.product?.id
        )

        return if (existingItem != null) {
            existingItem.amount += shoppingCart.amount
            updateProduct(existingItem)
            println("Cantidad de producto actualizada: ${existingItem.amount}")
            return ResponseEntity(Message("Cantidad de producto actualizada"), HttpStatus.OK)
        } else {
            addNewProduct(shoppingCart)
            println("Producto agregado al carrito: ${shoppingCart.product?.name} - Cantidad: ${shoppingCart.amount}")
            return ResponseEntity(Message("Producto agregado al carrito"), HttpStatus.OK)
        }
    }

    private fun addNewProduct(shoppingCart: ShoppingCart) {
        shoppingCartRepository.save(shoppingCart)
    }

    fun updateProduct(shoppingCart: ShoppingCart) {
        shoppingCartRepository.save(shoppingCart)
    }

    fun getCountByClient(clientId: Long?): Long {
        return shoppingCartRepository.countByClient_Id(clientId)
    }

    fun updateByID(request: ShoppingCart, id: Long?): ShoppingCart? {
        val shoppingCart = id?.let { shoppingCartRepository.findById(it).orElse(null) }
        if (shoppingCart != null) {
            shoppingCart.amount = request.amount
            shoppingCart.client = request.client
            shoppingCart.product = request.product
            shoppingCartRepository.save(shoppingCart)
        }
        return shoppingCart
    }

    fun findById(id: Long?): ShoppingCart? {
        return id?.let { shoppingCartRepository.findById(it).orElse(null) }
    }
}
